File: faunadb-java/src/main/java/com/faunadb/client/types/LazyValue.java
package com.faunadb.client.types;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.google.common.base.Optional;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;

import java.time.Instant;
import java.time.LocalDate;

import static java.lang.String.format;

/**
 * A {@link Value} that wraps a JSON response tree. This Value does not convert to a concrete type until one of its
 * type coercion methods is called.
 */
@JsonDeserialize(using = Codec.LazyValueDeserializer.class)
public final class LazyValue extends Value {

  private final JsonNode underlying;
  private final ObjectMapper json;

  private final Value rawJson = new Value() {
    @Override
    public String toString() {
      return underlying.toString();
    }
  };
  private Value lazy = rawJson;

  @JsonCreator
  LazyValue(JsonNode underlying, ObjectMapper json) {
    this.underlying = underlying;
    this.json = json;
  }

  @JsonValue
  private JsonNode underlying() {
    return underlying;
  }

  @Override
  public String asString() {
    return as(StringV.class).asString();
  }

  @Override
  public Optional<String> asStringOption() {
    return as(StringV.class).asStringOption();
  }

  @Override
  public Boolean asBoolean() {
    return as(BooleanV.class).asBoolean();
  }

  @Override
  public Optional<Boolean> asBooleanOption() {
    return as(BooleanV.class).asBooleanOption();
  }

  @Override
  public Long asLong() {
    return as(LongV.class).asLong();
  }

  @Override
  public Optional<Long> asLongOption() {
    return as(LongV.class).asLongOption();
  }

  @Override
  public Double asDouble() {
    return as(DoubleV.class).asDouble();
  }

  @Override
  public Optional<Double> asDoubleOption() {
    return as(DoubleV.class).asDoubleOption();
  }

  @Override
  public Instant asTs() {
    return as(TsV.class).asTs();
  }

  @Override
  public Optional<Instant> asTsOption() {
    return as(TsV.class).asTsOption();
  }

  @Override
  public LocalDate asDate() {
    return as(DateV.class).asDate();
  }

  @Override
  public Optional<LocalDate> asDateOption() {
    return as(DateV.class).asDateOption();
  }

  @Override
  public ImmutableList<Value> asArray() {
    return as(ArrayV.class).asArray();
  }

  @Override
  public Optional<ImmutableList<Value>> asArrayOption() {
    return as(ArrayV.class).asArrayOption();
  }

  @Override
  public ImmutableMap<String, Value> asObject() {
    return as(ObjectV.class).asObject();
  }

  @Override
  public Optional<ImmutableMap<String, Value>> asObjectOption() {
    return as(ObjectV.class).asObjectOption();
  }

  @Override
  public Ref asRef() {
    return as(Ref.class).asRef();
  }

  @Override
  public Optional<Ref> asRefOption() {
    return as(Ref.class).asRefOption();
  }

  @Override
  public SetRef asSetRef() {
    return as(SetRef.class).asSetRef();
  }

  @Override
  public Optional<SetRef> asSetRefOption() {
    return as(SetRef.class).asSetRefOption();
  }

  @Override
  public Value get(String key) {
    return asObject().get(key);
  }

  @Override
  public Value get(String... keys) {
    Value res = this;
    for (String key : keys)
      res = res.get(key);

    return res;
  }

  @Override
  public Optional<Value> getOption(String key) {
    Optional<ImmutableMap<String, Value>> object = asObjectOption();
    if (object.isPresent())
      return Optional.fromNullable(object.get().get(key));

    return Optional.absent();
  }

  @Override
  public Optional<Value> getOption(String... keys) {
    Optional<Value> res = Optional.<Value>of(this);
    for (String key : keys) {
      res = res.get().getOption(key);
      if (!res.isPresent()) break;
    }

    return res;
  }

  @Override
  public Value get(int index) {
    return asArray().get(index);
  }

  @Override
  public Optional<Value> getOption(int index) {
    Optional<ImmutableList<Value>> array = asArrayOption();

    if (array.isPresent()) {
      try {
        return Optional.of(array.get().get(index));
      } catch (IndexOutOfBoundsException ign) {
        return Optional.absent();
      }
    }

    return Optional.absent();
  }

  private Value as(Class<? extends Value> clazz) {
    if (lazy != rawJson)
      return lazy;

    try {
      lazy = json.convertValue(underlying, clazz);
    } catch (IllegalArgumentException ex) {
      //Failed to convert. Try again with other type.
    }

    return lazy;
  }

  @Override
  public boolean equals(Object other) {
    return other != null &&
      other instanceof LazyValue
      && underlying.equals(((LazyValue) other).underlying);
  }

  @Override
  public int hashCode() {
    return underlying.hashCode();
  }

  @Override
  public String toString() {
    return format("LazyValue(%s)", underlying.toString());
  }

}

File: faunadb-scala/src/main/scala/faunadb/types/LazyValue.scala
package faunadb.types

import com.fasterxml.jackson.databind.`type`.TypeFactory
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.{JsonNode, ObjectMapper}
import faunadb.response._
import faunadb.query._

import scala.collection.AbstractMap

/**
  * A [[Value]] that wraps a JSON response tree. This Value does not convert into a concrete
  * type until one of the type coercion methods is called.
  */
@JsonDeserialize(using=classOf[LazyValueDeserializer])
class LazyValue private[faunadb] (private val underlying: JsonNode, json: ObjectMapper) extends Value {
  def asBefore = Before(RawV(underlying))
  def asAfter = After(RawV(underlying))

  override def asStringOpt: Option[String] = if (underlying.isTextual) Some(underlying.asText) else None
  override def asBooleanOpt: Option[Boolean] = if (underlying.isBoolean) Some(underlying.asBoolean()) else None
  override def asNumberOpt: Option[Long] = if (underlying.isNumber) Some(underlying.asLong()) else None
  override def asDoubleOpt: Option[Double] = if (underlying.isDouble) Some(underlying.asDouble()) else None

  override def asArrayOpt: Option[Array[Value]] = {
    try {
      Option(json.convertValue(underlying, TypeFactory.defaultInstance().constructArrayType(classOf[LazyValue])))
    } catch {
      case _: IllegalArgumentException => None
    }
  }
  override def asObjectOpt: Option[collection.Map[String, Value]] = {
    try {
      Option(json.convertValue(underlying, classOf[LazyValueMap]))
    } catch {
      case _: ClassCastException => None
      case _: IllegalArgumentException => None
    }
  }

  override def asRefOpt: Option[Ref] = {
    try {
      Option(json.convertValue(underlying, classOf[Ref]))
    } catch {
      case _: IllegalArgumentException => None
    }
  }

  override def asPageOpt: Option[Page] ={
    try {
      Option(json.convertValue(underlying, classOf[Page]))
    } catch {
      case _: IllegalArgumentException => None
    }
  }

  override def asInstanceOpt: Option[Instance] = {
    try {
      Option(json.convertValue(underlying, classOf[Instance]))
    } catch {
      case _: IllegalArgumentException => None
    }
  }

  override def asKeyOpt: Option[Key] = {
    try {
      Option(json.convertValue(underlying, classOf[Key]))
    } catch {
      case _: IllegalArgumentException => None
    }
  }

  override def asDatabaseOpt: Option[Database] = {
    try {
      Option(json.convertValue(underlying, classOf[Database]))
    } catch {
      case _: IllegalArgumentException => None
    }
  }

  override def asTokenOpt: Option[Token] = {
    try {
      Option(json.convertValue(underlying, classOf[Token]))
    } catch {
      case _: IllegalArgumentException => None
    }
  }

  override def asClassOpt: Option[Class] = {
    try {
      Option(json.convertValue(underlying, classOf[Class]))
    } catch {
      case _: IllegalArgumentException => None
    }
  }

  override def asIndexOpt: Option[Index] = {
    try {
      Option(json.convertValue(underlying, classOf[Index]))
    } catch {
      case _: IllegalArgumentException => None
    }
  }

  override def asEventOpt: Option[Event] = {
    try {
      Option(json.convertValue(underlying, classOf[Event]))
    } catch {
      case _: IllegalArgumentException => None
    }
  }

  override def asSetOpt: Option[Set] = {
    try {
      Option(json.convertValue(underlying, classOf[Set]))
    } catch {
      case _: IllegalArgumentException => None
    }
  }

  override def asTsOpt: Option[Ts] = {
    try {
      Option(json.convertValue(underlying, classOf[Ts]))
    } catch {
      case ex: IllegalArgumentException =>
        println(underlying)
        println(ex)
        None
    }
  }

  override def asDateOpt: Option[Date] = {
    try {
      Option(json.convertValue(underlying, classOf[Date]))
    } catch {
      case _: IllegalArgumentException => None
    }
  }

  override def toString: String = underlying.toString

  override def equals(obj: scala.Any): Boolean = {
    obj.isInstanceOf[LazyValue] && underlying.equals(obj.asInstanceOf[LazyValue].underlying)
  }

  override def hashCode(): Int = underlying.hashCode()
}

/**
 * An immutable dictionary of response nodes. FaunaDB responses can be polymorphic, so this
 * dictionary allows individual entries to be coerced into concrete response types as required.
 *
 * ===Example===
 *
 * {{{
 *   val response: ResponseNode // Some result from the FaunaDB client
 *   val responseMap = response.asObject
 *   responseMap("someKey").asString
 *   responseMap("someNumberKey").asNumber
 *
 *   // these can be chained
 *   response.asObject("someObjectKey").asObject("someNestedKey").asString
 * }}}
 */
@JsonDeserialize(using=classOf[LazyValueMapDeserializer])
class LazyValueMap private[faunadb] (underlying: Map[String, LazyValue]) extends AbstractMap[String, LazyValue] {
  override def get(key: String): Option[LazyValue] = underlying.get(key)

  override def +[B1 >: LazyValue](kv: (String, B1)): collection.Map[String, B1] = underlying + kv

  override def iterator: Iterator[(String, LazyValue)] = underlying.iterator

  override def -(key: String): collection.Map[String, LazyValue] = underlying - key

  override def hashCode(): Int = underlying.hashCode()

  override def equals(that: Any): Boolean = underlying.equals(that)
}
